mayastor/tests/error_count_retry.rs

```diff
@@ -0,0 +1,187 @@
+    Reactor::block_on(async {
+        inject_error(
+            EE_ERROR_DEVICE,
+            SPDK_BDEV_IO_TYPE_WRITE,
+            VBDEV_IO_FAILURE,
+            1,
+        );
+        err_write_nexus(true).await;
+    });
+
+    reactor_run_millis(10); // give time for any errors to be added to the error store
+
+    nexus_err_query_and_test(
+        BDEV_EE_ERROR_DEVICE,
+        NexusErrStore::WRITE_FLAG,
+        1,
+        Some(1_000_000_000),
+    );
+
+    // 2 errors injected, 2 attempts allowed, 1 read attempt, 2 read errors
+    // should be logged and the IO should fail
+    Reactor::block_on(async {
+        inject_error(
+            EE_ERROR_DEVICE,
+            SPDK_BDEV_IO_TYPE_READ,
+            VBDEV_IO_FAILURE,
+            2,
+        );
+        err_read_nexus(false).await;
+    });
+
+    // IO should now succeed
+    Reactor::block_on(async {
+        err_read_nexus(true).await;
+    });
+
+    mayastor_env_stop(0);
+
+    common::delete_file(&[DISKNAME1.to_string()]);
+    common::delete_file(&[YAML_CONFIG_FILE.to_string()]);
+}
+
+async fn create_nexus() {
+    let ch = vec![BDEV_EE_ERROR_DEVICE.to_string()];
+
+    nexus_create(ERROR_COUNT_TEST_NEXUS, 64 * 1024 * 1024, None, &ch)
+        .await
+        .unwrap();
+}
+
+fn nexus_err_query_and_test(
+    child_bdev: &str,
+    io_type_flags: u32,
+    expected_count: u32,
+    age_nano: Option<u64>,
+) {
+    let nexus = nexus_lookup(ERROR_COUNT_TEST_NEXUS).unwrap();
+    let count = nexus
+        .error_record_query(
+            child_bdev,
+            io_type_flags,
+            NexusErrStore::IO_FAILED_FLAG,
+            age_nano,
+            QueryType::Total,
+        )
+        .expect("failed to query child");
+    assert!(count.is_some()); // true if the error_store is enabled
+    assert_eq!(count.unwrap(), expected_count);
+}
+
+async fn err_write_nexus(succeed: bool) {
+    let bdev = Bdev::lookup_by_name(ERROR_COUNT_TEST_NEXUS)
+        .expect("failed to lookup nexus");
+    let d = bdev
+        .open(true)
+        .expect("failed open bdev")
+        .into_handle()
+        .unwrap();
+    let buf = d.dma_malloc(512).expect("failed to allocate buffer");
+
+    match d.write_at(0, &buf).await {
+        Ok(_) => {
+            assert_eq!(succeed, true);
+        }
+        Err(_) => {
+            assert_eq!(succeed, false);
+        }
+    };
+}
+
+async fn err_read_nexus(succeed: bool) {
+    let bdev = Bdev::lookup_by_name(ERROR_COUNT_TEST_NEXUS)
+        .expect("failed to lookup nexus");
+    let d = bdev
+        .open(true)
+        .expect("failed open bdev")
+        .into_handle()
+        .unwrap();
+    let mut buf = d.dma_malloc(512).expect("failed to allocate buffer");
+
+    match d.read_at(0, &mut buf).await {
+        Ok(_) => {
+            assert_eq!(succeed, true);
+        }
+        Err(_) => {
+            assert_eq!(succeed, false);
+        }
+    };
+}
+
+fn reactor_run_millis(milliseconds: u64) {
```

Comment:
IIRC you've used this kind of function before, maybe worth moving it into common?

Reply:
True, will fix.